common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java

```diff
@@ -0,0 +1,240 @@
+    private int pageDestination = 1;
+    private List<Integer> customJumpDestinations;
+
+    @SubscribeEvent
+    public void onScreenInit(ScreenInitEvent e) {
+        if (!(e.getScreen() instanceof AbstractContainerScreen<?> screen)) return;
+        if (!Models.Container.isBankScreen(screen)) return;
+
+        isBankScreen = true;
+
+        containerId = ((AbstractContainerScreen<?>) e.getScreen()).getMenu().containerId;
+
+        currentPage = Models.Container.getCurrentBankPage(screen);
+
+        customJumpDestinations = List.of(
+                buttonOnePage.get(),
+                buttonTwoPage.get(),
+                buttonThreePage.get(),
+                buttonFourPage.get(),
+                buttonFivePage.get(),
+                buttonSixPage.get());
+
+        if (!quickJumping) return;
+
+        if (pageDestination > lastPage) {
+            quickJumping = false;
+            pageDestination = currentPage;
+        } else if (pageDestination != currentPage) {
+            jumpToDestination();
+        } else {
+            quickJumping = false;
+        }
+    }
+
+    @SubscribeEvent
+    public void onContainerClose(ContainerCloseEvent.Post e) {
+        isBankScreen = false;
+        currentPage = 1;
+        pageDestination = 1;
+    }
+
+    @SubscribeEvent
+    public void onSlotClicked(ContainerClickEvent e) {
+        if (!isBankScreen) return;
+
+        int slotIndex = e.getSlotNum();
+
+        if (BUTTON_SLOTS.contains(slotIndex)) {
+            int buttonIndex = BUTTON_SLOTS.indexOf(slotIndex);
+            pageDestination = customJumpDestinations.get(buttonIndex);
+            e.setCanceled(true);
+            jumpToDestination();
+        }
+    }
+
+    @SubscribeEvent
+    public void onSetSlot(SetSlotEvent.Pre e) {
+        if (!isBankScreen) return;
+        if (e.getContainer() instanceof Inventory) return;
+
+        if (BUTTON_SLOTS.contains(e.getSlot())) {
+            ItemStack jumpButton = new ItemStack(Items.DIAMOND_AXE);
+            jumpButton.setDamageValue(92);
+
+            CompoundTag jumpTag = jumpButton.getOrCreateTag();
+            jumpTag.putInt("HideFlags", 6);
+            jumpTag.putBoolean("Unbreakable", true);
+            jumpButton.setTag(jumpTag);
+
+            int buttonIndex = BUTTON_SLOTS.indexOf(e.getSlot());
+            int buttonDestination = customJumpDestinations.get(buttonIndex);
+
+            jumpButton.setCount(buttonDestination);
+
+            jumpButton.setHoverName(Component.literal(ChatFormatting.GRAY + "Jump to Page " + buttonDestination));
+
+            e.setItemStack(jumpButton);
+        }
+    }
+
+    @SubscribeEvent
+    public void onRenderSlot(SlotRenderEvent.Pre e) {
+        if (!(e.getScreen() instanceof AbstractContainerScreen<?> screen)) return;
+        if (!isBankScreen) return;
+
+        if (e.getSlot().index == NEXT_PAGE_SLOT && Models.Container.isLastBankPage(screen)) {
+            lastPage = currentPage;
```

Comment:
This is all that is left of the rendering. I think we can move it to a more suitable location. onSetSlot should work fine, I think..?

Reply:
I did try that, but wasn't able to get a reference to the screen to pass in to isLastBankPage. Any idea on how to get it?

Reply:
The ContainerModel is a bit wonky designed to work with screens... The best solution would probably be to look for e.g. MenuEvent.MenuOpenedEvent and check getTitle(). This requires rewriting the ContainerModel functionality a bit.

But since this is even more work, I am totally fine with keeping this for now. We can clean this up later as well, if/when we fix the ContainerModel.

Reply:
Ok, I'll add a comment there as a reminder and then it should be good to go and I'll work on the misc bucket

Reply:
I did realize though that a possibly better solution than the SetSlotEvent is to use ContainerSetContentEvent.Pre, where you get all the items at once, and can just replace the individual item stacks. You will not get a container name, but you do get a containerId which you can use to verify that it is the correct container. (But for this to work, you will need to save the containerId, perhaps from a MenuOpenedEvent.)

You have done enough of work on this, so I fully understand if you want to pass on doing this. Let me know if that is the case, and I'll approve it as it is. But if you strive for perfection (and a spot in the Wynntils dev team? :-)), please go ahead!

Reply:
Did try using ContainerSetContentEvent.Pre yesterday after Mcrtin suggested it and the items didn't render but I'll give it another try quick

Reply:
Have a look at how `ItemHandler.onContainerSetContent` works...

Reply:
I'm definitely close, it renders them once I go to a different page but not the first time the container opens.

Reply:
This is what I've got, I'm not sure why it doesn't work on the first open, I checked the hovername of the items after the loop by calling e.getItems() again and they did have the new names but when it goes to render it's still using the old items.

```
@SubscribeEvent
    public void onContainerSetContent(ContainerSetContentEvent.Pre e) {
        if (e.getContainerId() != containerId) return;

        List<ItemStack> items = e.getItems();

        for (int buttonSlot : BUTTON_SLOTS) {
            ItemStack jumpButton = new ItemStack(Items.DIAMOND_AXE);
            jumpButton.setDamageValue(92);

            CompoundTag jumpTag = jumpButton.getOrCreateTag();
            jumpTag.putInt("HideFlags", 6);
            jumpTag.putBoolean("Unbreakable", true);
            jumpButton.setTag(jumpTag);

            int buttonIndex = BUTTON_SLOTS.indexOf(buttonSlot);
            int buttonDestination = customJumpDestinations.get(buttonIndex);

            jumpButton.setCount(buttonDestination);

            jumpButton.setHoverName(Component.literal(ChatFormatting.GRAY + "Jump to Page " + buttonDestination));

            items.set(buttonSlot, jumpButton);
        }
    }
```

Reply:
Hm, I can't say straight up what's wrong. Can you check by debugging if you even end up in the method for the first page? Is the containerId correct, so you actually go on to change the items, even on the first page?

Do you get SetSlotEvent afterwards (that is how it is usually sent; don't know why)? Can you check if you get that for the first page also? If you do, can you see if you have gotten your new itemstack or if the old one is still in place?

Reply:
Did do a bit of debugging, I called e.getItems after the loop and checked the hovernames of items as those indexes and it was updated correctly. Then in SetSlot I think it had the old ItemStack so hadn't updated there, but then it does work for every other page which is strange

Reply:
This is really strange. I noted recently that ContainerSetContentEvent.Pre is called twice every time you open a new container, and then only once. Maybe this is related somehow... Otherwise I don't know. In worst case, you'll have to drop this change and go back to the per-slot setting.

Reply:
Yeah I did notice that when the bank first opens it's called once then every other page it's called twice. I wasn't sure if that was something on the servers end or was actually to do with the event. Well someone figures it out in the future can always come back and retry it but for now guess it'll have to stay in SetSlot

Reply:
In the code above, you seem to have a `containerId` field. Is it possible you can use that part of the code, to check if it is the last bank screen, instead of the render function? If we skip the ContainerSetContentEvent part.

Reply:
I was getting that in onScreenInit, and I don't think the items have rendered by that point so it won't be able to check the name of the item. I can double check though.

Also, just as a quick fyi in case you were about to merge this, I've found a potential issue whilst making this for the misc bucket where if you try to jump to your final page which is 1 below one of the hard coded jump pages, it won't jump there. So I'm just adding a bit more to tryUsingJumpButtons

Reply:
Actually I'm not sure if the method I was thinking of will work. 

Aside from a render event constantly checking the title of the container to see if it's where we're supposed to be, is there a way to check if after we've tried to click a slot that it has actually moved page?

Also, calling isLastBankPage inside onScreenInit didn't work,

Reply:
I've fixed this by adding the last pages to storage and in `tryUsingJumpButtons` when it calculates the closest jump, in won't allow it if the jump is greater than the last page.

I've added this method to `ContainerModel` which also eliminates the need for checking for last page in the render event. As this is called in `onScreenInit`.

```
public int updateFinalPage(Screen screen) {
        int currentPage = getCurrentBankPage(screen);

        if (isBankScreen(screen)) {
            if (currentPage > finalBankPage.get()) {
                finalBankPage.store(currentPage);
            }

            return finalBankPage.get();
        } else if (isBookshelfScreen(screen)) {
            if (currentPage > finalBookshelfPage.get()) {
                finalBookshelfPage.store(currentPage);
            }

            return finalBookshelfPage.get();
        } else if (isMiscBucketScreen(screen)) {
            if (currentPage > finalMiscBucketPage.get()) {
                finalMiscBucketPage.store(currentPage);
            }

            return finalMiscBucketPage.get();
        }

        return currentPage;
    }
```

The only problem with this, is that because I've set the default final pages to their maximum, until you manually visit the final page, the issue of not jumping to your final page if it's one before the hard coded jumps will still not work, but I think this is the best solution anyway.

Reply:
Ok, I decided to go to the bottom of this. `isLastBankPage` is just a bit stupidly implemented for our case, since it tries to do too much.

The core thing it does is
```
StyledText.fromComponent(cs.getMenu().getSlot(8).getItem().getHoverName())
                        .endsWith(LAST_BANK_PAGE_STRING);
```
and this you could test from the full container update event.

So my suggestion is that you basically extract this part into a separate method in ContainerModel, like `isItemIndicatingLastBankPage`, and move the 8 to a public constant `LAST_BANK_PAGE_SLOT`. Then call this new function from the old isLastBankPage(), and also use it in your new code.

Reply:
Should all be done now, I'll just make the changes on my misc bucket branch and check the jumping to last page works there and this should be good to go

Reply:
Yup, seems to work there so once you visit your final page once it'll all work fine, hopefully most people will figure that out and we won't get too many complaints of people thinking the feature is broken